[SPARK-13904][Scheduler]Add support for pluggable cluster manager

hbhanawat opened the pull request:
## What changes were proposed in this pull request?

This commit adds support for pluggable cluster manager. And also allows a cluster manager to clean up tasks without taking the parent process down.

To plug a new external cluster manager, ExternalClusterManager trait should be implemented. It returns task scheduler and backend scheduler that will be used by SparkContext to schedule tasks. An external cluster manager is registered using the java.util.ServiceLoader mechanism (This mechanism is also being used to register data sources like parquet, json, jdbc etc.). This allows auto-loading implementations of ExternalClusterManager interface.

Currently, when a driver fails, executors exit using system.exit. This does not bode well for cluster managers that would like to reuse the parent process of an executor. Hence,
1. Moving system.exit to a function that can be overriden in subclasses of CoarseGrainedExecutorBackend.
2. Added functionality of killing all the running tasks in an executor.
## How was this patch tested?

ExternalClusterManagerSuite.scala was added to test this patch.

Commit:
This commit adds support for pluggable cluster manager. And also allows a cluster manager to clean up tasks without taking the parent process down.

To plug a new external cluster manager, ExternalClusterManager trait should be implemented. It returns task scheduler and backend scheduler that will be used by SparkContext to schedule tasks. An external cluster manager is registered using the java.util.ServiceLoader mechanism (This mechanism is also being used to register data sources like parquet, json, jdbc etc.). This allows auto-loading implementations of ExternalClusterManager interface.

Currently, when a driver fails, executors exit using system.exit. This does not bode well for cluster managers that would like to reuse the parent process of an executor. Hence,

  1. Moving system.exit to a function that can be overriden in subclasses of CoarseGrainedExecutorBackend.
  2. Added functionality of killing all the running tasks in an executor.

AmplabJenkins commented:
Can one of the admins verify this patch?

sarutak commented:
ok to test.

SparkQA commented:
**[Test build #53264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53264/consoleFull)** for PR 11723 at commit [`800834f`](https://github.com/apache/spark/commit/800834f24ad1f0c4a68d8d49f600db6570d100ef).

SparkQA commented:
**[Test build #53264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53264/consoleFull)** for PR 11723 at commit [`800834f`](https://github.com/apache/spark/commit/800834f24ad1f0c4a68d8d49f600db6570d100ef).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExternalClusterManager `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53264/
Test FAILed.

rxin commented:
Hmm I'm not sure if it makes sense to create a public API for this at this point. This is directly exposing something that is very internal to the current implementation of Spark (SchedulerBackend), and these are by no means stable.

hbhanawat commented:
@rxin  Thanks for commenting. 

Spark was designed such that it is agnostic to the underlying cluster manager (as long as it can acquire executor processes, and these communicate with each other). Since Spark is now being used in newer and different use cases, there is a need for allowing other cluster managers to manage spark components. One such use case is - embedding spark components like executor and driver inside another process which may be a datastore. This allows co-location of data and processing. Another use case would be using Spark like an application server (you might have heard about spark-jobserver). Spark's current design allows handling such use cases if the cluster manager supports it. Hence, IMO, it is meaningful to allow plugging in new cluster managers. 

From code perspective, I think that even creation of TaskScheduler and SchedulerBackend for Yarn/Mesos/local mode should be done using a similar interface.

rxin commented:
@hbhanawat I understand that. The problem is not whether you can find a single legitimate use case.

The introduction of every API always benefit something -- there is no argument about it. Otherwise nobody would be adding new APIs. The question is how many it benefits, and how the APIs can be maintained and evolved. You are effectively taking a bunch of private APIs that were never meant to be public and making them public. This approach is not maintainable.

rxin commented:
@hbhanawat to be clear, I think we might be able to add this as a semi-private API and external resource managers can use, but with the understanding that this is tied to specific versions of Spark and might break from release over release. Exposing this as a stable public API based on my experience seeing how Spark has evolved in the last 5 years is not going to work.

hbhanawat commented:
@rxin ok, I get it. I would make ExternalClusterManager as private[spark] and mark it as developer API. I hope that should suffice.

Commit:
Fixed issues in the test

SparkQA commented:
**[Test build #53424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53424/consoleFull)** for PR 11723 at commit [`e87c1e0`](https://github.com/apache/spark/commit/e87c1e020046796f2b55bd3d50d10eebca6a8470).

SparkQA commented:
**[Test build #53424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53424/consoleFull)** for PR 11723 at commit [`e87c1e0`](https://github.com/apache/spark/commit/e87c1e020046796f2b55bd3d50d10eebca6a8470).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53424/
Test PASSed.

Commit:
Since the scheduler APIs are not stable and may change in future, marking ExternalClusterManager as DeveloperApi and making it as private[spark].

SparkQA commented:
**[Test build #53525 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53525/consoleFull)** for PR 11723 at commit [`ae808d7`](https://github.com/apache/spark/commit/ae808d73e022077dba6ad999627589eed4730270).

SparkQA commented:
**[Test build #53525 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53525/consoleFull)** for PR 11723 at commit [`ae808d7`](https://github.com/apache/spark/commit/ae808d73e022077dba6ad999627589eed4730270).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53525/
Test PASSed.

hbhanawat commented:
@rxin I have completed the changes. Please review.

hbhanawat commented:
@rxin Any update? Any changes needed from my side?

tejasapatil commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
+      // exactly one registered manager
+      case head :: Nil => Some(head)
+      case Nil => None
+      case multipleMgrs => sys.error(s"Multiple Cluster Managers registered " +
```
Can you include the list of matching cluster managers in the message ?

hbhanawat replied:
Done.

tejasapatil commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -149,7 +149,14 @@ private[spark] class Executor(
     }
   }
-
+  def killAllTasks (interruptThread: Boolean) : Unit = {
```
add a space between the methods

tejasapatil replied:
I could not see this method being called from anywhere. If you don't plan to use it, please remove this.

hbhanawat replied:
1. Done 
2. We are targeting a use case where in executors are launched inside another running process and when an executor goes down, it does not take the parent parent process down. In such cases, executors should kill the running tasks when they go down.  Given that, the runningtasks a private val, we need a method that can be called from the executor backend.

tejasapatil commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -149,7 +149,14 @@ private[spark] class Executor(
+        taskRunner.kill(interruptThread)
+      }
+    }
+  }
```
add a space between the methods

hbhanawat replied:
Done

tejasapatil commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,62 @@
+/**
+ * :: DeveloperApi ::
+ * A cluster manager interface to plugin external scheduler.
+ *
```
remove extra space

hbhanawat replied:
Done.

tejasapatil commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,65 @@
+  }
+}
+
+class CheckExternalClusterManager extends ExternalClusterManager {
```
Rename to `TestExternalClusterManager` or `DummyExternalClusterManager`

hbhanawat replied:
Done.

tejasapatil commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,62 @@
+   * @param sc SparkContext
+   * @return TaskScheduler that will be responsible for task handling
+   */
+  def createTaskScheduler (sc: SparkContext): TaskScheduler
```
can you also include the `masterURL` as a param ?

hbhanawat replied:
I have added the master url. In the hindsight, I think it makes sense.

tejasapatil commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,62 @@
+   * @param scheduler TaskScheduler that will be used with the scheduler backend.
+   * @return SchedulerBackend that works with a TaskScheduler
+   */
+  def createSchedulerBackend (sc: SparkContext, scheduler: TaskScheduler): SchedulerBackend
```
can you also include the `masterURL` as a param ?

hbhanawat replied:
Done.

tejasapatil commented on core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager:
```diff
@@ -0,0 +1 @@
+org.apache.spark.scheduler.CheckExternalClusterManager
```
why is this file needed ?

hbhanawat replied:
To instantiate the dummy cluster manager in the test.

tejasapatil commented on dev/.rat-excludes:
```diff
@@ -98,3 +98,4 @@ LZ4BlockInputStream.java
 spark-deps-.*
 .*csv
 .*tsv
+org.apache.spark.scheduler.ExternalClusterManager
```
why is this needed ?

hbhanawat replied:
Rat throws an error for the files that do not have apache license. Hence excluding this file.

tejasapatil replied:
The version of `ExternalClusterManager` in the PR has a license header in it. Can you remove this and try ?

hbhanawat replied:
I believe there is some confusion.  There is a file named org.apache.spark.scheduler.ExternalClusterManager in a test folder that is used by the ServiceLoader to launch the dummy external cluster manager. This file does not (and cannot) have a header and hence it needs to be excluded.

tejasapatil commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,65 @@
+
+}
+
+class FakeScheduler extends TaskScheduler {
```
To keep this consistent with the external CM declared above, you could rename this to `TestTaskScheduler` or `DummyTaskScheduler`

tejasapatil commented:
@rxin : I would really like to have this PR in trunk. As things stand, for anyone using their own scheduler, one has to maintain a patch over open source release to have that glue in Spark. Re API breaking over Spark releases : I agree that breaking APIs is bad. But it would be atleast better than the current model of dealing with this: doing a merge for *every* release.

rxin commented:
Is this something Facebook needs too?

tejasapatil commented:
@rxin : Yes !! At Facebook we are using an internal scheduler to run Spark executors. Maintaining an internal patch to have that "glue" and merging it against every Spark release can be avoided by this PR. Ideally we want to get to a place where there are no patches to maintain and everything is just pluggable. If we get there, testing RCs would be easier and we can help flagging issues early on (especially ones related to scalability).

rxin commented:
OK - once you are done with your own review ping me. I will take a look at it again.

Commit:
Added review comments.
1. Fixed formatting issues
2. Added master url as part of the other functions of ExternalClusterManager

SparkQA commented:
**[Test build #54037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54037/consoleFull)** for PR 11723 at commit [`59990bf`](https://github.com/apache/spark/commit/59990bffad47eeab20435c2ba34590f92fafabfc).

tejasapatil commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,67 @@
+        setAppName("testcm").set("spark.driver.allowMultipleContexts", "true")
+    sc = new SparkContext(conf)
+    // check if the scheduler components are created
+    assert(sc.schedulerBackend.isInstanceOf[FakeSchedulerBackend])
```
Sorry I missed this in the last review comments yesterday. I thought that `FakeSchedulerBackend` was in this same file and you could rename it but now I see that its from some other place. 

While reading, it feels odd to have Fake* and then Dummy* test classes. I am not sure about the whats followed in Spark codebase. Couple options:
- rename Dummy* classes => Fake*. Move all the Fake* classes to a common test utils file for the module.
- Instead of re-using `FakeSchedulerBackend` from another place, create a `FakeSchedulerBackend` here.

hbhanawat replied:
I too  missed it completely.  
I think it wasn't a great idea in the first place to use FakeSchedulerBackend of some other class from maintenance perspective. I am going ahead with your option 2.

tejasapatil commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,67 @@
+  }
+}
+
+class DummyExternalClusterManager extends ExternalClusterManager {
```
`private`

hbhanawat replied:
Done.

tejasapatil commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,67 @@
+
+}
+
+class DummyTaskScheduler extends TaskScheduler {
```
`private`

hbhanawat replied:
Done

SparkQA commented:
**[Test build #54037 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54037/consoleFull)** for PR 11723 at commit [`59990bf`](https://github.com/apache/spark/commit/59990bffad47eeab20435c2ba34590f92fafabfc).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54037/
Test FAILed.

Commit:
Fixed class names and their scope

SparkQA commented:
**[Test build #54066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54066/consoleFull)** for PR 11723 at commit [`1517f17`](https://github.com/apache/spark/commit/1517f17b5e7bac94e3d1f73c85c2553891f9defa).

hbhanawat commented:
@rxin  @tejasapatil  the previous build failure is not related to my checkin  and looks like some other issue as other build requests failed with the same exception. In case, it fails again, how do I restart it?

SparkQA commented:
**[Test build #2682 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2682/consoleFull)** for PR 11723 at commit [`1517f17`](https://github.com/apache/spark/commit/1517f17b5e7bac94e3d1f73c85c2553891f9defa).

SparkQA commented:
**[Test build #54066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54066/consoleFull)** for PR 11723 at commit [`1517f17`](https://github.com/apache/spark/commit/1517f17b5e7bac94e3d1f73c85c2553891f9defa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54066/
Test PASSed.

SparkQA commented:
**[Test build #2682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2682/consoleFull)** for PR 11723 at commit [`1517f17`](https://github.com/apache/spark/commit/1517f17b5e7bac94e3d1f73c85c2553891f9defa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

hbhanawat commented:
@tejasapatil if you are done with the review, can we ask @rxin to have a look at this?

tejasapatil commented:
sure. @rxin can you please review the PR ?

tejasapatil commented:
Can someone review this PR ?

hbhanawat commented:
@rxin can you please review this PR?

rxin commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -135,6 +135,8 @@ private[spark] class CoarseGrainedExecutorBackend(
     }
   }
+
+  def exitExecutor(): Unit = System.exit(1)
```
can you add some document here saying why this is a public API - otherwise somebody might come in and remove this next time this code gets used.

rxin commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -150,6 +150,15 @@ private[spark] class Executor(
     }
   }
 
+  def killAllTasks (interruptThread: Boolean) : Unit = {
```
same here -- say why this is a public thing and how it is used.

also remove the space after killAllTasks

rxin commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,65 @@
+  * A cluster manager interface to plugin external scheduler.
+  */
+@DeveloperApi
+private[spark] trait ExternalClusterManager {
```
are you sure you want this to be private spark?

rxin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
+          cm.initialize(scheduler, backend)
+          (backend, scheduler)
+        } catch {
+          case e: Exception => {
```
catch NonFatal here, i.e.
```
case NonFatal(e) =>
```

rxin commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
+  }
+
+  private def getClusterManager(url: String): Option[ExternalClusterManager] = {
+    val loader = Utils.getContextOrSparkClassLoader
```
maybe
```scala
val loader = Utils.getContextOrSparkClassLoader
val serviceLoaders = ServiceLoader.load(classOf[ExternalClusterManager], loader).asScala.filter(_.canCreate(url))
if (serviceLoaders.size > 1) {
  throw new SparkException("Multiple Cluster Managers ($multipleMgrs) registered for the URL $url")
}
serviceLoaders.headOption
```

rxin commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,65 @@
+private[spark] trait ExternalClusterManager {
+
+  /**
+    * Check if this cluster manager instance can create scheduler components
```
use javadoc style indent in spark, i.e.

```
/**
 * Check if this cluster manager instance can create scheduler components
 * ...
 */
```

rxin commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,65 @@
+    * @param scheduler TaskScheduler that will be used with the scheduler backend.
+    * @return SchedulerBackend that works with a TaskScheduler
+    */
+  def createSchedulerBackend(sc: SparkContext,
```
the way we indent is
```
def createSchedulerBackend(
    sc: SparkContext,
    masterURL: String,
    scheduler: TaskScheduler): SchedulerBackend
```

rxin commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,74 @@
+  override def start(): Unit = {}
+  override def stop(): Unit = {}
+  override def submitTasks(taskSet: TaskSet): Unit = {}
+  override def cancelTasks(stageId: Int, interruptThread: Boolean) {}
```
add a Unit here for return type to be consistent

rxin commented on core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala:
```diff
@@ -0,0 +1,74 @@
+
+  def canCreate(masterURL: String): Boolean = masterURL == "myclusterManager"
+
+  def createTaskScheduler(sc: SparkContext,
```
can you fix the indent for the functions?

rxin commented:
Sorry for the delay. This looks pretty good. Just have some comments about the style to be more consistent with rest of the Spark codebase.

Commit:
Fixed code style issues.

SparkQA commented:
**[Test build #55994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55994/consoleFull)** for PR 11723 at commit [`696cc71`](https://github.com/apache/spark/commit/696cc715b3cf43cf10300057b00193fee601ec45).

Commit:
Merge branch 'master' into pluggableScheduler

Conflicts:
	core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

rxin commented:
there are some conflicts with master - can you rebase? Thanks.

rxin commented:
The latest changes LGTM

SparkQA commented:
**[Test build #55997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55997/consoleFull)** for PR 11723 at commit [`2a81517`](https://github.com/apache/spark/commit/2a8151713739ecd1da867f3ddd75d436fe1ef101).

SparkQA commented:
**[Test build #55997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55997/consoleFull)** for PR 11723 at commit [`2a81517`](https://github.com/apache/spark/commit/2a8151713739ecd1da867f3ddd75d436fe1ef101).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55997/
Test FAILed.

Commit:
Fixed scala style issues

SparkQA commented:
**[Test build #55998 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55998/consoleFull)** for PR 11723 at commit [`6747420`](https://github.com/apache/spark/commit/674742029e28278b8ff038590464045f714b5790).

SparkQA commented:
**[Test build #55998 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55998/consoleFull)** for PR 11723 at commit [`6747420`](https://github.com/apache/spark/commit/674742029e28278b8ff038590464045f714b5790).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55998/
Test FAILed.

SparkQA commented:
**[Test build #55994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55994/consoleFull)** for PR 11723 at commit [`696cc71`](https://github.com/apache/spark/commit/696cc715b3cf43cf10300057b00193fee601ec45).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExternalClusterManager `

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55994/
Test PASSed.

hbhanawat commented:
test this please

hbhanawat commented:
@rxin how do I get this retested by Jenkins? There were few issues going on with the Jenkins when I checked in my last changes and now it is not retesting it?

srowen commented:
Jenkins retest this please

srowen commented:
Jenkins add to whitelist

SparkQA commented:
**[Test build #56012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56012/consoleFull)** for PR 11723 at commit [`6747420`](https://github.com/apache/spark/commit/674742029e28278b8ff038590464045f714b5790).

SparkQA commented:
**[Test build #56012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56012/consoleFull)** for PR 11723 at commit [`6747420`](https://github.com/apache/spark/commit/674742029e28278b8ff038590464045f714b5790).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56012/
Test PASSed.

rxin commented on core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala:
```diff
@@ -0,0 +1,65 @@
+ * A cluster manager interface to plugin external scheduler.
+ */
+@DeveloperApi
+trait ExternalClusterManager {
```
i just realized most of the return types used in this class are private[spark], so your implementation of this interface would need to be in the spark package anyway. I'm going to add private[spark] to this when I merge.

rxin commented:
Merging in master. Thanks.

rxin commented:
One thing - can you guys try to see if you can implement one of the existing cluster manager with this, and then we can make sure this is a proper API? Otherwise it is really easy to get removed because it is currently unused by anything in Spark.

hbhanawat commented:
@rxin I will open another JIRA and a PR to do this. Thanks for the review.

tedyu commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -140,6 +140,13 @@ private[spark] class CoarseGrainedExecutorBackend(
+   * executor exits differently. For e.g. when an executor goes down,
+   * back-end may not want to take the parent process down.
+   */
+  protected def exitExecutor(): Unit = System.exit(1)
```
Should a parameter of exit code (int) be added to this method ?

tedyu replied:
Created https://github.com/apache/spark/pull/12457
